overlord/aspectstate/aspectstate.go

```diff
@@ -0,0 +1,108 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspectstate
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/aspects"
+	"github.com/snapcore/snapd/overlord/aspectstate/aspecttest"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func Set(st *state.State, account, directory, aspect, field, value string) error {
+	st.Lock()
+	defer st.Unlock()
+
+	if err := aspecttest.MaybeMockAspect(st); err != nil {
+		return err
+	}
+
+	var asps map[string]map[string]*aspects.Directory
```

Comment:
I don't think we want to marshal entire directories, only the data

Reply:
Changed it to serialize the databag only and re-create the aspect directory based on the patterns we get from the assertion (which for now is just a list of patterns)